pandas/core/indexes/numeric.py

```diff
@@ -123,6 +125,21 @@ def insert(self, loc, item):
             item = self._na_value
         return super().insert(loc, item)
 
+    def _union(self, other, sort):
+        # float | [u]int -> float
+        # <T>   | <T>    -> T
+        # <T>   | <U>    -> object
+        needs_cast = (
+            (is_integer_dtype(self.dtype) and is_float_dtype(other.dtype)) or
+            (is_integer_dtype(other.dtype) and is_float_dtype(self.dtype))
+        )
+        if needs_cast:
+            first = self.astype("float")
```

Comment:
maybe float64 here?

Reply:
These are both indexes, so right now that's the same.

In a future with 32-bit indexes, I think `.astype('float')` is better, since Int32Index.astype('float') would be Float32Index?